tempesta_fw/sched/tfw_sched_ratio.c

```diff
@@ -0,0 +1,467 @@
+		recalc |= tfw_apm_stats(sl->srvs[s].srv->apm, &pstats);
+		weight[s] = pstats.val[idx] ? : 1;
+	}
+	if (!recalc)
+		return 1;
+
+	tfw_sched_ratio_calc(weight, ratio, sl->srv_n);
+
+	return 0;
+}
+
+/*
+ * Calculate ratios for servers with equal weights.
+ * Return 1 to indicate that all ratios/weights are equal.
+ */
+static int
+tfw_sched_ratio_equal(TfwSrvGroup *sg, int *ratio)
+{
+	size_t s;
+	TfwRatioSrvList *sl = sg->sched_data;
+	int eqshare = TFW_RATIO_SCALE / sl->srv_n ? : 1;
+
+	for (s = 0; s < sl->srv_n; ++s)
+		ratio[s] = eqshare;
+
+	return 1;
+}
+
+/*
+ * Calculate ratios for servers with static, possibly different weights.
+ * Return 1 to indicate that all ratios/weights are equal.
+ * Return 0 to indicate that ratios are calculated by individial weights.
+ */
+static int
+tfw_sched_ratio_static(TfwSrvGroup *sg, int *ratio)
+{
+	size_t s;
+	TfwRatioSrvList *sl = sg->sched_data;
+	int diff = 0, wequal, weight[sl->srv_n];
+
+	wequal = sl->srvs[0].srv->weight;
+	for (s = 0; s < sl->srv_n; ++s) {
+		weight[s] = sl->srvs[s].srv->weight;
+		diff |= (wequal != weight[s]);
+	}
+	if (!diff) {
+		tfw_sched_ratio_equal(sg, ratio);
+		return 1;
+	}
+
+	tfw_sched_ratio_calc(weight, ratio, sl->srv_n);
+
+	return 0;
+}
+
+/**
+ * Add a connection and a server, if new, to the scheduler.
+ * Called at configuration stage, no synchronization is required.
+ */
+static void
+tfw_sched_ratio_add_conn(TfwSrvGroup *sg, TfwServer *srv, TfwConnection *conn)
+{
+	size_t s, c;
+	TfwRatioSrv *cl;
+	TfwRatioSrvList *sl = sg->sched_data;
+
+	BUG_ON(!sl);
+
+	for (s = 0; s < sl->srv_n; ++s)
+		if (sl->srvs[s].srv == srv)
+			break;
+	BUG_ON(s == TFW_SG_MAX_SRV);
+
+	cl = &sl->srvs[s];
+
+	if (s == sl->srv_n) {
+		int nodiff, ratio[sl->srv_n + 1];
+		TfwRatioRated *rated = sl->rated;
+
+		cl->srv = srv;
+		++sl->srv_n;
+
+		if (sg->flags & TFW_SG_F_SCHED_RATIO_STATIC)
+			nodiff = tfw_sched_ratio_static(sg, ratio);
+		else if (sg->flags & TFW_SG_F_SCHED_RATIO_DYNAMIC)
+			nodiff = tfw_sched_ratio_equal(sg, ratio);
+		else
+			BUG();
+
+		if (nodiff)
+			tfw_sched_fill_equal(sg, ratio, rated);
+		else
+			tfw_sched_fill_rated(sg, ratio, rated);
+	}
+
+	for (c = 0; c < cl->conn_n; ++c) {
+		if (cl->conns[c] == conn) {
+			TFW_WARN("sched=[%s]: attempt to add an existing "
+				 "connection: sg=[%s] srv=[%zd] conn=[%zd]\n",
+				 sg->sched->name, sg->name, s, c);
+			return;
+		}
+	}
+	BUG_ON(c == TFW_SRV_MAX_CONN);
+
+	cl->conns[c] = conn;
+	++cl->conn_n;
+}
+
+/**
+ * On each subsequent call the function returns the next available
+ * connection to one of the servers in the group. Connections to a
+ * server are rotated in pure round-robin manner. The server is chosen
+ * according to its current weight. Servers with more weight are chosen
+ * more often than servers with less weight. 
+ * Dead connections and servers w/o live connections are skipped.
+ */
+static TfwConnection *
+tfw_sched_ratio_get_srv_conn(TfwMsg *msg, TfwSrvGroup *sg)
```

Comment:
You have asked about other ways to implement scheduler, here is my suggestion. Add `int repeats; atomic64_t s_counter;` to `TfwRatioSrv`. Choosing server will be done in this way:
```
for (s = 0; s < sl->srv_n; ++s) {
    i = (uint64_t)atomic64_inc_return(&sl->counter) % sl->srv_n;
    srv_cl = &sl->srvs[i];

    /*  
     * Make less than srv_cl->repeats requests to one server in a row, 
     * and then move to next one 
     */
    uint64_t continue_with_server =  
        (uint64_t)atomic64_inc_return(& srv_cl->s_counter) %  (srv_cl->repeats + 1);
    it (!continue_with_server)
       continue;
    
    /* Get connection for server. */
    ...
}
return NULL;
```
That seems to be naive, but that does the trick. I think with this approach we don't need to have rcu or make `TfwRatioSrv->repeats` atomic, since it updated quite rare.